Move rpc calls to REST

kannappanr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
In distributed mode, REST API will be used to acquire and manage locks instead
of RPC.

RPC has been completely removed from MinIO source.

Since we are moving from RPC to REST, rolling upgrades cannot be used, as the
nodes that have not yet been upgraded cannot talk to the ones that have
been upgraded.

We expect all the MinIO processes on all nodes to be stopped and then the
upgrade process to be completed.

## Motivation and Context
Moving to REST API instead of RPC 

## Regression
<!-- Is this PR fixing a regression? (Yes / No) -->
No
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
Using `mc` to copy a big folder to minio running in distributed mode.
`mc admin top locks` command can be used to verify that this works.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

kannappanr committed:
Lock: Move rpc calls to REST

harshavardhana commented on cmd/lock-rest-client.go:
```diff
@@ -0,0 +1,316 @@
+// RLock calls read lock REST API.
+func (client *lockRESTClient) RLock(args dsync.LockArgs) (reply bool, err error) {
+
+	var reader bytes.Buffer
```
you should borrow the current code of restricting the RLock bytes.Buffer to 1K buffer since without allocation bytes.Buffer grows at 2x per byte.

kannappanr replied:
done

harshavardhana reviewed (changes requested):
Also since this changes the locking behavior, how does it affect rolling upgrades?

harshavardhana committed:
remove rpc

harshavardhana commented on cmd/lock-rest-client.go:
```diff
@@ -0,0 +1,316 @@
+	restClient, err := rest.NewClient(serverURL, tlsConfig, rest.DefaultRESTTimeout, newAuthToken)
+
+	if err != nil {
+		return &lockRESTClient{serverURL: serverURL, host: peer, restClient: restClient, connected: false}, err
```
If this is returned upon error and not used, can we not just return nil here?

kannappanr replied:
done

harshavardhana reviewed (changes requested):
Can you elaborate the commit message body a bit more, with more details regarding the change and how the rolling upgrade is not possible with this?

kannappanr committed:
Lock: Use REST API instead of RPC

In distributed mode, use REST API to acquire and manage locks instead
of RPC.

RPC has been completely removed from MinIO source.

Since we are moving from RPC to REST, we cannot use rolling upgrades as the
nodes that have not yet been upgraded cannot talk to the ones that have
been upgraded.

We expect all the minio processes on all nodes to be stopped and then the
upgrade process to be completed.

krishnasrinivas commented on cmd/lock-rest-client.go:
```diff
@@ -0,0 +1,316 @@
+// after verifying format.json
+func (client *lockRESTClient) call(method string, values url.Values, body io.Reader, length int64) (respBody io.ReadCloser, err error) {
+	if !client.connected {
+		err := client.reConnect()
```
need to implement exponential backoff, else tcp connect to down servers everytime might be expensive

harshavardhana committed:
Force http1.1 for internode communication

kannappanr committed:
Modify RestClient constructor

kannappanr committed:
Add exponential backoff retry logic

kannappanr committed:
Fix review comments
